### DISCUSSION TITLE
redux/state management

### DISCUSSION THREAD

**Discussion by dave-doty** [pyedifice/pyedifice#102] [Q&A]
I think of React+Redux (as opposed to React alone) as the closest Javascript gets to The Elm Architecture/MVU. Edifice appears to do the React part fairly closely, but seems to dispense with the Redux/Elm concept of a single global Model/State object, instead using React state in the examples, which is a much more distributed notion of State/Model from the MVU than Redux global state, which (in my opinion) couples the View structure to the Model much more closely than in Elm or Redux+React.

Is there a straightforward way/recommendation to incorporate Redux (e.g., [ReduxPY](https://github.com/Carsten-Leue/ReduxPY)) or some similar state management with Edifice? In particular, I appreciate not having to think about how to distribute the model throughout the view, or having to refactor to move React state further up the view tree when it needs to be shared across multiple components.

In Elm, in a view function (representing a view component), I just access the part of the Model it needs, and Elm itself goes about figuring out how to connect the Model to the View, i.e., how to re-render that component only when there's a change in the part of the Model it accesses.

In React/Redux, I have to be a bit more explicit in setting up binding between the two (in particular creating one or more ["connected components"](https://react-redux.js.org/api/connect) that explicitly connect Redux state/model fields to React prop fields). Nevertheless, once this is done, it still decouples the update part from the "how the model gets distributed among the various components of the view". For instance, if a user clicks a button and that means some part of the model should change that affect various other pieces of the view, the code that responds to the button click simply dispatches an Redux Action to change the Model, and the reducer handling that Action (implementation of the update part of MVU within Redux) updates the Model, also without having to think about which parts of the View depend on that piece of the Model.

**Comment by jamesdbrock** [comment]
Thanks very much for your thoughtful post @dave-doty .

> Edifice appears to do the React part fairly closely, but seems to dispense with the Redux/Elm concept of a single global Model/State object

Yes.

> Is there a straightforward way/recommendation to incorporate Redux (e.g., [ReduxPY](https://github.com/Carsten-Leue/ReduxPY)) or some similar state management with Edifice?

No. Thanks for pointing me to ReduxPY, I hadn't seen that before.

I'm sympathetic to the idea that Edifice should provide more API help in structuring apps with large global state. That's the kind of app that I'm trying to write at work, and I too would like help from Edifice.

But I'm not yet sure how it should be done.

As Dan Abramov points out in [You Might Not Need Redux](https://medium.com/@dan_abramov/you-might-not-need-redux-be46360cf367), you can easily write Redux in terms of `useState`, but it may not be worth it “*unless you have a plan* to benefit from this additional indirection.”

**Comment by jamesdbrock** [comment]
Maybe what we need is a Hook that is like [`useContext`](https://legacy.reactjs.org/docs/hooks-reference.html#usecontext).

**Reply by xaviergmail** [reply]
I came here looking for a way to achieve some sort of global state (after realizing the pyedifice.org docs were out of date and `StateValue` is deprecated) and while I don't have a use case for Redux specifically, `useContext` would be fantastic!

**Reply by jamesdbrock** [reply]
Hi @xaviergmail . Yeah, `StateValue` and `StateManager` (in [the old docs](https://www.pyedifice.org/state) ) was David Deng’s attempt at reactive global state. David implemented `subscribe(component)` but he never implemented the most important part, which is *un*subscribing when a component is unmounted. So `StateValue` and `StateManager` were never completely working. So I deprecated them.

Hm yeah maybe we should have `useContext`.

**Reply by jamesdbrock** [reply]
https://github.com/pyedifice/pyedifice/issues/104

**Reply by dave-doty** [reply]
I have never been very familiar with `useContext`. After reading up quite a bit on it after seeing this, it seems like it can replace Redux in certain scenarios, but is far from being able to handle all the same things: https://blog.isquaredsoftware.com/2021/01/context-redux-differences/ I'm particularly concerned about performance issues. In my application, re-rendering the whole thing on every state update is way too slow, but it seems that simply trying to use one large global State as "context" available to the entire View tree would have this effect. (unless I'm misunderstanding something) 

Making it more performant seems to require something like breaking up the State into multiple contexts in different parts of the View tree, and then it seems we're back to the same problem of coupling the shape of the State data tightly to the shape of the View tree. (again, unless I'm misunderstanding something)

Beyond potential performance issues with using a single global Context, [here](https://blog.isquaredsoftware.com/2021/01/context-redux-differences/#when-should-i-use-redux-instead) it lists when Redux remains superior to `useContext` (and all of these apply to my case):

- You have larger amounts of application state that are needed in many places in the app.
- The app state is updated frequently over time.
- The logic to update that state may be complex.
- The app has a medium or large-sized codebase, and might be worked on by many people.
- You want to be able to understand when, why, and how the state in your application has updated, and visualize the changes to your state over time.
- You need more powerful capabilities for managing side effects, persistence, and data serialization.

So even if there's no plan to implement a Redux/Edifice connect module, I'm still curious if the authors have a suggestion how I might implement my own on top of Edifice, since I still prefer the React+Redux way of doing things to React+Context.

**Reply by jamesdbrock** [reply]
> re-rendering the whole thing on every state update is way too slow, but it seems that simply trying to use one large global State as "context" available to the entire View tree would have this effect.

Yeah I think that's right.

In Edifice (and React) you can skip rendering part of the tree by separating that part of the tree into a `component`. Then pass props to the `component`. If the props don't change, the `component` will not be re-rendered.

https://github.com/pyedifice/pyedifice/blob/07f57947c65cdcd2ef36f29288089a47ecd9be80/edifice/engine.py#L593-L597

But if every component depends on one big global state, then every component will be re-rendered with every change to that state.

I have an app with a big global state. I pass small pieces of that state as props to components. I also pass updater functions as props to the components which they can call to update the global state.

These four points I think we would get from `useContext`.

> - You have larger amounts of application state that are needed in many places in the app.
> - The app state is updated frequently over time.
> - The logic to update that state may be complex.
> - The app has a medium or large-sized codebase, and might be worked on by many people.

These two points we would only get if we built out a whole suite of tools like https://redux-toolkit.js.org/ . That is what Dan Abramov means by [“benefit from this additional indirection.”](https://medium.com/@dan_abramov/you-might-not-need-redux-be46360cf367) An Edifice-Toolkit would be great, but I don't personally want to do that work right now. PRs welcome.

> - You want to be able to understand when, why, and how the state in your application has updated, and visualize the changes to your state over time.
> - You need more powerful capabilities for managing side effects, persistence, and data serialization.

**Reply by jamesdbrock** [reply]
Oh, I read the [article](https://blog.isquaredsoftware.com/2021/01/context-redux-differences/) and I see more what you're talking about @dave-doty 

> Context + `useReducer` relies on passing the *current state value* via Context. React-Redux passes the *current Redux store instance* via Context.
> That means that when `useReducer` produces a new state value, *all* components that are subscribed to that context will be forced to re-render, even if they only care about *part* of the data. This *may* lead to performances issues, depending on the size of the state value, how many components are subscribed to that data, and how often they re-render. With React-Redux, components can subscribe to specific pieces of the store state, and only re-render when those values change.

So Redux has a thing where you subscribe to updates on only a part of a global state.

**Comment by dave-doty** [comment]
> I have an app with a big global state. I pass small pieces of that state as props to components. I also pass updater functions as props to the components which they can call to update the global state.

This is something that `useContext` handles nicely? Like, do you make a separate context object for each component, and pull the relevant pieces from the global state into that context? Sorry I'm naive, but just about every article I find about `useContext` is some tiny example where the whole state is one or two variables, and the whole View is a tiny little tree, and the context is the whole state. So it shows the syntax of `useContext` but not in a way that scales.

I haven't been able to find an example of `useContext` that actually discusses (1) performance and avoiding re-renders by using minimal state into each context object (if indeed that's the intended usage of `useContext` in an app with a large View that needs to access many different parts of the global state), and (2) how to decouple the shape of the global application State from the shape of the View tree. The state in these examples is always some React state, meaning it's still coupling application State tightly to the View (though I thought the whole point of any MV* architecture is to separate the M from the V).

If you do implement `useContext` in #104, it would be fantastic if it also came with an example of recommended usage, which is not one of these tiny, non-scalable examples. :)

**Comment by jamesdbrock** [comment]
> This is something that `useContext` handles nicely? 

No. The [article](https://blog.isquaredsoftware.com/2021/01/context-redux-differences/) puts it well: “the primary purpose for using Context is to avoid prop-drilling” for read-only global state like theme or language.

> it's still coupling application State tightly to the View (though I thought the whole point of any MV* architecture is to separate the M from the V).

State is data. View is a pure function from State to Widget tree. They are tightly coupled, as they should be. The point of MVU is that it allows us to write the View function, which is relatively simple, and then the framework does the work of diffing the resulting Widget trees. The basic mechanism for *skipping the work* of re-rendering the whole tree is component props. Redux also supplies an alternate “subscription” mechanism for skipping that work. Which would be nice-to-have, but not essential.

**Comment by jamesdbrock** [comment]
Thanks for raising these issues @dave-doty , you have clarified my understanding of Redux.

**Comment by xaviergmail** [comment]
Just chiming in here (without having read the entire conversation)to provide more _context_ 😉 

While entirely unrelated to redux, a good example of something I'd like to use `use_context` for is passing a reference to a boto3 session. In ReactJS, this could be an Apollo client, Axios client, site theme (light/dark), etc. It simplifies the development process of views tremendously.

For example, I have an AWS profile switcher dropdown which provides aioboto3 session and components that need to interact with AWS either to populate fields or perform actions can just use the session obtained from the context anywhere in the tree underneath the provider. Components in different branches of the view tree should have separate values for the context.

A little draft:

```python
DEFAULT_SESSION = aioboto3.Session()
SessionContext = Context(DEFAULT_SESSION)

@component
def ProfileSwitcher(self):
    available, set_available = use_state(["default"])
    use_effect(lambda: set_available(SessionContext.available_profiles), [])

    profile, set_profile = use_state("default")

    session, set_session = use_state(DEFAULT_SESSION)
    use_effect(lambda profile: set_session(aioboto3.Session(profile_name=profile)), [profile])

    with View(layout="column"):
        with View(layout="row"):
            Label("AWS Profile:")
            Dropdown(
                selection=profile,
                options=available,
                on_select=set_profile,
            )

        with SessionContext.provide(session):
            yield # Currently not possible within @component? Unless I'm missing something?


@ed.component
def AccountIdLabel(self):
    # Assume we're nested down several components here, kept short for brevity
    session = use_context(SessionContext)
    account_id, set_account_id = use_state("loading...")

    async def load_account_id(session):
        async with session.client('sts') as client:
            # Edit: Initializing a client performs significant IO, we could instead have the ProfileSwitcher provide a bigger context object containing initialized clients/resources 
            identity = await client.get_caller_identity()
            set_account_id(identity['Account'])

    use_async(load_account_id, [session])

    ed.Label(f"Account ID: {account_id}")

@component
def Main(self):
    with View(layout="column"):
        with ProfileSwitcher():
            AccountIdLabel()
        with ProfileSwitcher():
            AccountIdLabel()
        AccountIdLabel()
```

In this example, assuming the different aws profiles are different accounts, we should see two profile selectors, each with a label containing that context's session's account id, independent from one another; followed by a lone account id label that constantly shows the default profile's account id (since it's not the child of any context providers)

**Comment by jamesdbrock** [comment]
UPDATE 2024-12-28

Edifice has `use_context` published in [v2.9.0](https://pyedifice.github.io/versions.html#v2-9-0)

**Comment by jamesdbrock** [comment]
I have published Edifice v3.1.0 with [`use_context_select`](https://pyedifice.github.io/stubs/edifice.use_context_select.html).
